README.md
# CRISPRCasFinder

CRISPRCasFinder is an updated, improved, and integrated version of CRISPRFinder and CasFinder. 
![CRISPR-Cas++](crispr-cas_logo.png)

# References/Citations

If you use this software, please cite: 

- Grissa I, Vergnaud G, Pourcel C. CRISPRFinder: a web tool to identify clustered regularly interspaced short palindromic repeats. <b>Nucleic Acids Res.</b> 2007 Jul;35(Web Server issue):W52-7. DOI: https://doi.org/10.1093/nar/gkm360 [PMID:17537822](https://www.ncbi.nlm.nih.gov/pubmed/17537822)

- Abby SS, Néron B, Ménager H, Touchon M, Rocha EP. MacSyFinder: a program to mine genomes for molecular systems with an application to CRISPR-Cas systems. <b>PLoS One.</b> 2014 Oct 17;9(10):e110726. DOI: https://doi.org/10.1371/journal.pone.0110726 [PMID:25330359](https://www.ncbi.nlm.nih.gov/pubmed/25330359)

- Couvin D, Bernheim A, Toffano-Nioche C, Touchon M, Michalik J, Néron B, Rocha EPC, Vergnaud G, Gautheret D, Pourcel C.
CRISPRCasFinder, an update of CRISRFinder, includes a portable version, enhanced performance and integrates search for Cas proteins.
<b>Nucleic Acids Res.</b> 2018 Jul 2;46(W1):W246-W251. DOI: https://doi.org/10.1093/nar/gky425 [PMID:29790974](https://www.ncbi.nlm.nih.gov/pubmed/29790974)

- Néron B, Denise R, Coluzzi C, Touchon M, Rocha EPC, Abby SS (2022). MacSyFinder v2: Improved modelling and search engine to identify molecular systems in genomes. <b>bioRxiv</b> DOI: 10.1101/2022.09.02.506364 [Link](https://www.biorxiv.org/content/10.1101/2022.09.02.506364v1)

Further information are available at: https://crisprcas.i2bc.paris-saclay.fr.

# Quick Installation
## Conda/Bioconda/Mamba
Note that you will first need to install [conda/bioconda](http://www.ddocent.com/bioconda/) to run the following commands:
```bash
conda env create -f ccf.environment.yml -n crisprcasfinder
conda activate crisprcasfinder
mamba init
mamba activate
mamba install -c bioconda macsyfinder=2.0
macsydata install -u CASFinder==3.1.0
```

## MacOS
```bash
./installer_MAC.sh
```

## Ubuntu
```bash
bash installer_UBUNTU.sh
source ~/.profile
```

## CentOS
Please first install conda if it is not already installed:
```bash
sudo yum -y update
sudo yum -y upgrade
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
export PATH=/path/to/miniconda3/bin/:$PATH
source ~/.bashrc
conda init
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
```
Now you can install CRISPRCasFinder as follows:
```bash
bash installer_CENTOS.sh
exit #this command could be needed if your command prompt changes
source ~/.bashrc
```

## Fedora
```bash
sudo yum -y update
sudo yum -y upgrade
bash installer_FEDORA.sh
source ~/.bashrc
```

You can run the command 'perl CRISPRCasFinder.pl -v' to see if everything is OK.
You may need to reinstall some Perl's modules (with command: sudo cpanm ...), for example: "sudo cpanm Date::Calc".
The notification "Possible precedence issue with control flow operator ..." will not affect results of analysis.
For further information, please see the documentation.

## To run CRISPRCasFinder in the current directory with example sequence you can type:
```bash
perl CRISPRCasFinder.pl -in install_test/sequence.fasta -cas -keep
```
For further details, please see the documentation.

# Documentation
A more complete User Manual is available at the following file : CRISPRCasFinder_Viewer_manual.pdf

# Licence

[GPL v3](https://github.com/dcouvin/CRISPRCasFinder/blob/master/COPYING)

# Container

If you want to try CRISPRCasFinder without installing dependencies,
The standalone version is also available as a singularity container (hosted on the [Download page of the CRISPR-Cas++ portal](https://crisprcas.i2bc.paris-saclay.fr/Home/Download)):

<!--[![https://www.singularity-hub.org/static/img/hosted-singularity--hub-%23e32929.svg](https://www.singularity-hub.org/static/img/hosted-singularity--hub-%23e32929.svg)](https://singularity-hub.org/collections/1624)-->

## To run the container
### Former version of CRISPRCasFinder (v4.2.20)
After downloading the [CrisprCasFinder.simg](https://crisprcas.i2bc.paris-saclay.fr/Home/DownloadFile?filename=CrisprCasFinder.simg) image from the [CRISPR-Cas++ Download page](https://crisprcas.i2bc.paris-saclay.fr/Home/Download), you can for example run the following command (sequence.fasta file must be replaced by your file):

```bash
singularity exec -B $PWD CrisprCasFinder.simg perl /usr/local/CRISPRCasFinder/CRISPRCasFinder.pl -so /usr/local/CRISPRCasFinder/sel392v2.so -cf /usr/local/CRISPRCasFinder/CasFinder-2.0.3 -drpt /usr/local/CRISPRCasFinder/supplementary_files/repeatDirection.tsv -rpts /usr/local/CRISPRCasFinder/supplementary_files/Repeat_List.csv -cas -def G -out RES21092020_2 -in sequence.fasta
```
<!--
```bash
singularity run shub://dcouvin/CRISPRCasFinder:4.2.18 -def General -cas -i my_sequence.fasta -keep
```
or download the image locally, and optionally rename it, then run it
```bash
singularity pull --name CRISPRCasFinder shub://dcouvin/CRISPRCasFinder:4.2.18 
./CRISPRCasFinder -def General -cas -i my_sequence.fasta -keep
```
-->

Please visit the following link for more information about singularity containers: https://www.sylabs.io/docs/

# Outline of the CRISPRCasFinder workflow

<img src="http://www.pasteur-guadeloupe.fr/files/Workflow_CRISPRCasFinder.png" title="CRISPRCasFinder workflow"></img>


NEWS.txt
## News from latest version of CRISPRCasFinder (version 4.3.2)

# Change #1
# CRISPR arrays having one to three spacers will now be rated with evidence-level 4 if their consensus repeat matched with a level 4 CRISPR array's consensus repeat (from the same studied sequence).
# The new definition for evidence-levels is now:
# - Level 1 indicates CRISPR-like arrays having 3 spacers or less,
# - Level 2 indicates CRISPR arrays having an entropy-based conservation (EBcons) of repeats lower than 70.
# - Level 3 indicates CRISPR arrays having a EBcons of repeats greater or equal to 70, and a spacer conservation (BioPerl’s overall percentage identity) greater than 8%,
# - Level 4 indicates CRISPR arrays having a EBcons of repeats greater or equal to 70, and a spacer conservation (BioPerl’s overall percentage identity) lower than 8%. Please note that CRISPR-like arrays having 3 spacers or less will be rated with level 4 if their consensus repeat matched with a formal level 4 CRISPR array's consensus repeat.

# Change #2  
# In CRISPRCasFinder version 4.2.19, MacSyFinder will not be launched to search Cas proteins if no CDS was detected.
# This improvement allows to avoid MacSyFinder's "RuntimeError".

# Change #3
# Installation procedures have been added for CentOS and Fedora operating systems.

# Change #4
# In CRISPRCasFinder version 4.2.20, the following modifications have been done:
# - The error concerning the analysis of multifasta file has been fixed. This error was due to a bad management of the launching of MacSyFinder/CasFinder 
#   when no CDS was detected in a sequence found in the multifasta.
# - CasFinder version 2.0.3 (the most recent updated version of CasFinder) has been added to the repository.
#   Please also note that path names to Cas definitions and profiles have been modified in order to fix the issue concerning detection of cas genes.
#   The former CasFinder-2.0.2 folder has been removed.
# - Please note that the "-rcfowce" option used in the online version of CRISPRCasFinder may produce different results as compared to the command line version of the tool, 
#   because this option ("-rcfowce") allows users to run CasFinder only when a CRISPR locus has been detected. 
#   The version of the dependent software (such as EMBOSS) may also cause slight differences.
#   The command line used during an analysis can be visualized in the result JSON file.
# - Please note that the Singularity image corresponding to the current version of the tool has not yet been updated.
# - Please note that this version of CasFinder (2.0.3) does not correspond to the last update 
#   (Makarova et al., 2020 ; https://doi.org/10.1038/s41579-019-0299-x). This change will be performed soon thanks to colleagues from 
#   Institut Pasteur in Paris. Thank you for your patience.

# Change #5
# Please note that a new singularity container is available and  hosted at the Download page of the CRISPR-Cas++ portal (https://crisprcas.i2bc.paris-saclay.fr/Home/Download)

# Change #6
# Please note that in the former installer bash files, the path corresponding to MacSyFinder version 1.0.5 was: https://github.com/gem-pasteur/macsyfinder/archive/refs/tags/macsyfinder-1.0.5.tar.gz

# Change #7
# Please note that a new version of CRISPRCasFinder is in development (version 4.3.1) using MacSyFinder version 2.0 (https://github.com/gem-pasteur/macsyfinder) and CASFinder version 3.1.0.
# You can use conda/mamba to install the new version of MacSyFinder and corresponding CASFinder version.
# Example of command lines to install the new version of MacSyFinder/CASFinder:
conda install -c conda-forge mamba
mamba install -c bioconda macsyfinder=2.0
macsydata install -u CASFinder==3.1.0
conda install -c bioconda perl-bioperl-core
# Please use a conda environment when running the code (several tests are in progress)
# Note that following Perl modules are no more needed: JSON::Parse, Date::Calc, Unix::Sysexits
# Please note that you will need to first install conda and activate it to run the new bash installer files
# Note that former CasFinder folder ("CasFinder-2.0.3") is no more needed in CRISPRCasFinder version 4.3.1

# Change #8
# Please note that some changes have been applied to CRISPRCasFinder (version 4.3.2). As version 4.3.1, this version also use MacSyFinder version 2.0
# Note that a new installation has been added to run the programme through a conda/mamba environment (instructions are provided in the README file)
# Note that CRISPRCasFinder vers 4.3.2 need MUSCLE version 5.1 in order to work properly. Otherwise, you will need to modify the CRISPRCasFinder.pl file
# indeed, the new version of MUSCLE can be run using the following command: "muscle -align <inputfile> -output <outputfile>"
# instead of the former command: "muscle -in <inputfile> -out <outputfile>"
# Furthermore, this new version of CCF still need "sel392v2.so" file to work (a Linux Ubuntu version of this file is provided in the GitHub repository),
# if needed, "sel392v2.so" file could be generated from other operating systems following Vmatch documentation/installation (http://www.vmatch.de/)



installer_FEDORA.sh
#!/bin/bash

# shell script allowing to install all CRISPRCasFinder.pl-v4.3's dependencies on Fedora
# (could be potentially adapted for Red Hat)
#
# same version than CRISPRCasFinder.pl, here 4.3.2
# authors: David Couvin, Fabrice Leclerc, Claire Toffano-Nioche

#------------------------
function launchInstall {
# $1 $packageManagmentInstall
# $2 package name
# $3 $LOGFILE
    echo "Installation of $2" >> $3
    $1 $2 >> $3
}
#------------------------

#save the current directory:
CURDIR=`pwd` 

#create log file:
LOGFILE=$CURDIR/install.log
if [ -e $LOGFILE ]; then rm $LOGFILE ; fi

sudo chmod 755 .
sudo chmod 755 *

#create src & bin folders:
echo "create $CURDIR/src and $CURDIR/bin folders" >> $LOGFILE
if [ ! -d $CURDIR/src ];then mkdir $CURDIR/src; fi
if [ ! -d $CURDIR/bin ];then mkdir $CURDIR/bin; fi

#geting OS info: 
ostype=`echo $OSTYPE`
echo "$ostype" >> $LOGFILE

if [ ! "$ostype" = "linux-gnu" ]; then
    echo 'Sorry, install process only for OSTYPE linux-gnu (based on yum)'
else
    echo "yum update/upgrade" >> $LOGFILE
    #sudo yum -y update >> $LOGFILE
    #sudo yum -y upgrade >> $LOGFILE
    packageManagmentInstall='sudo yum -y install '
    distribution=''

    if $(uname -m | grep '64'); then
       distribution='Linux_x86_64-64bit'
       echo "$ostype distribution is: 64-bit" >> $LOGFILE
    else
       distribution='Linux_i386-32bit'
       echo "$ostype distribution is: 32-bit" >> $LOGFILE
    fi

    #important packages
    if [ ! -x "$(command -v wget)" ]; then
        launchInstall "$packageManagmentInstall" "wget" "$LOGFILE"
    fi
    if [ ! -x "$(command -v curl)" ]; then
        launchInstall "$packageManagmentInstall" "curl" "$LOGFILE"
    fi
    if [ ! -x "$(command -v git)" ]; then
        launchInstall "$packageManagmentInstall" "git" "$LOGFILE"
    fi
    if [ ! -x "$(command -v java)" ]; then
        launchInstall "$packageManagmentInstall" "java" "$LOGFILE"
    fi
    if [ ! -x "$(command -v python2.7)" ]; then
        launchInstall "$packageManagmentInstall" "python-unversioned-command" "$LOGFILE"
    fi
    if [ ! -x "$(command -v parallel)" ]; then
        launchInstall "$packageManagmentInstall" "parallel" "$LOGFILE"
    fi
    if [ ! -x "$(command -v cpanm)" ]; then
        launchInstall "$packageManagmentInstall" "perl-App-cpanminus" "$LOGFILE"
    fi
    if [ ! -x "$(command -v cpan)" ]; then
        launchInstall "$packageManagmentInstall" "perl-CPAN" "$LOGFILE"
    fi
    launchInstall "$packageManagmentInstall" "perl-Module-Build" "$LOGFILE"
  
    #"bioinfo" packages
    if [ ! -x "$(command -v hmmsearch)" ]; then
        launchInstall "$packageManagmentInstall" "hmmer" "$LOGFILE"
        #launchInstall "$packageManagmentInstall" "hmmer-devel" "$LOGFILE"   # may not be necessary
    fi  
    if [ ! -x "$(command -v fuzznuc)" ] && [ ! -x "$(command -v needle)" ]
    then
        launchInstall "$packageManagmentInstall" "EMBOSS" "$LOGFILE"  
        #launchInstall "$packageManagmentInstall" "EMBOSS-devel" "$LOGFILE"  # may not be necessary
        launchInstall "$packageManagmentInstall" "EMBOSS-libs" "$LOGFILE"
    fi
	
    launchInstall "$packageManagmentInstall" "ncbi-blast+" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-Time-Piece" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-XML-Simple" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-Digest-MD5" "$LOGFILE"
    
    if [ ! -x "$(command -v gcc)" ]; then
        launchInstall "$packageManagmentInstall" "gcc-c++" "$LOGFILE"
    fi
    if [ ! -x "$(command -v make)" ]; then
        launchInstall "$packageManagmentInstall" "make" "$LOGFILE"
    fi

    #cpanm
    #sudo cpanm Unix::Sysexits >> $LOGFILE
    sudo cpanm Bio::Perl >> $LOGFILE
    sudo cpanm Bio::FeatureIO >> $LOGFILE
    #sudo cpanm Try::Tiny >> $LOGFILE
    #sudo cpanm Test::Most >> $LOGFILE
    #sudo cpanm JSON::Parse >> $LOGFILE
    sudo cpanm Class::Struct >> $LOGFILE
    sudo cpanm Bio::DB::Fasta >> $LOGFILE
    sudo cpanm File::Copy  >> $LOGFILE
    #sudo cpanm Bio::Seq Bio::SeqIO >> $LOGFILE
    #sudo cpanm --force Bio::Tools::Run::Alignment::Muscle >> $LOGFILE
    sudo cpanm Date::Calc >> $LOGFILE
    sudo cpanm Bio::AlignIO >> $LOGFILE
    sudo cpanm Getopt::Long >> $LOGFILE

    #BioPerl-Run
    # wget https://cpan.metacpan.org/authors/id/C/CJ/CJFIELDS/BioPerl-Run-1.007003.tar.gz 
    # most recent version of BioPerl-Run will not work. We should use release-1-7-0
    wget https://github.com/bioperl/bioperl-run/archive/release-1-7-0.tar.gz >> $LOGFILE
    tar -xzvf release-1-7-0.tar.gz >> $LOGFILE
    cd bioperl-run-release-1-7-0
    perl Build.PL
    sudo ./Build test
	sudo ./Build installdeps
    sudo ./Build install
    cd $CURDIR
	
    #install vmatch
    if [ ! -x "$(command -v vmatch2)" ] && [ ! -x "$(command -v mkvtree2)" ] && [ ! -x "$(command -v vsubseqselect2)" ]
    then
      echo "Installation of Vmatch" >> $LOGFILE
      echo "change directory to $CURDIR/src" >> $LOGFILE
      cd $CURDIR/src
      wget http://vmatch.de/distributions/vmatch-2.3.0-${distribution}.tar.gz >> $LOGFILE
      tar -zxf vmatch-2.3.0-${distribution}.tar.gz >> $LOGFILE
      gcc -Wall -Werror -fPIC -O3 -shared vmatch-2.3.0-${distribution}/SELECT/sel392.c -o $CURDIR/sel392v2.so >> $LOGFILE
      echo "copy $CURDIR/src/vmatch-2.3.0-${distribution}/vmatch, mkvtree and vsubseqselect to $CURDIR/bin/" >> $LOGFILE
      sudo cp $CURDIR/src/vmatch-2.3.0-${distribution}/vmatch $CURDIR/bin/vmatch
      sudo cp $CURDIR/src/vmatch-2.3.0-${distribution}/mkvtree $CURDIR/bin/mkvtree
      sudo cp $CURDIR/src/vmatch-2.3.0-${distribution}/vsubseqselect $CURDIR/bin/vsubseqselect
      echo "change directory to $CURDIR" >> $LOGFILE
      cd $CURDIR
    else
        echo "Vmatch2, mkvtree2, and vsubseqselect2 are already installed." >> $LOGFILE
    fi

    #install muscle
    if [ ! -x "$(command -v muscle)" ]; then
        echo "Installation of Muscle" >> $LOGFILE
        launchInstall "$packageManagmentInstall" "muscle" "$LOGFILE" # may not work properly
    else
        echo "Muscle is already installed." >> $LOGFILE
    fi
    
    if [ ! -x "$(command -v muscle)" ]; then
      echo "Installation of Muscle (second trial)" >> $LOGFILE
      if [ $distribution == 'Linux_x86_64-64bit' ]
      then
        wget http://drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux64.tar.gz >> $LOGFILE
        tar -xzvf muscle3.8.31_i86linux64.tar.gz >> $LOGFILE
        sudo chmod +x muscle3.8.31_i86linux64
        sudo cp muscle3.8.31_i86linux64 /usr/bin/muscle
      else 
        wget http://drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux32.tar.gz >> $LOGFILE
        tar -xzvf muscle3.8.31_i86linux32.tar.gz >> $LOGFILE
        sudo chmod +x muscle3.8.31_i86linux32
        sudo cp muscle3.8.31_i86linux32 /usr/bin/muscle
      fi
    else
        echo "Muscle is already installed." >> $LOGFILE
    fi

    #install prodigal
    if [ ! -x "$(command -v prodigal)" ] 
    then
        echo "Installation of Prodigal" >> $LOGFILE
        wget https://github.com/hyattpd/Prodigal/archive/GoogleImport.zip >> $LOGFILE
        unzip GoogleImport.zip
        cd Prodigal-GoogleImport
        sudo make install
    else
        echo "Prodigal is already installed." >> $LOGFILE
    fi

    #install macsyfinder
    if [ ! -x "$(command -v macsyfinder)" ] 
    then
      echo "Installation of MacSyFinder using conda/mamba" >> $LOGFILE
      conda install -c conda-forge mamba
      mamba install -c bioconda macsyfinder=2.0
      macsydata install -u CASFinder==3.1.0
      conda install -c bioconda perl-bioperl-core
      #cd ${CURDIR}
      #wget https://github.com/gem-pasteur/macsyfinder/archive/refs/tags/macsyfinder-1.0.5.tar.gz >> $LOGFILE
      #tar -xzf macsyfinder-1.0.5.tar.gz
      #test -d bin ||  mkdir bin
      #cd bin
      #ln -s ../macsyfinder-macsyfinder-1.0.5/bin/macsyfinder
      #cd ${CURDIR}
      #echo "add definition of MACSY_HOME (${CURDIR}/macsyfinder-1.0.5/) in .bashrc" >> $LOGFILE
      #echo "export MACSY_HOME=${CURDIR}/macsyfinder-1.0.5/" >> $HOME/.bashrc
      #echo "add bin folder ($CURDIR/bin) to the definition of PATH in $HOME/.bashrc" >> $LOGFILE
      #echo "export PATH=${CURDIR}/bin:${PATH}" >> $HOME/.bashrc
    else
      echo "MacSyFinder is already installed." >> $LOGFILE
    fi

    #set environment variables
    #source $HOME/.bashrc #this command must be typed directly by user
    
    echo "Installation done."
	echo "You may need to reinstall some Perl's modules (with command: sudo cpanm ...)"
	echo "Please type command 'source ~/.bashrc' in order to reload your .bashrc file, then run 'perl CRISPRCasFinder.pl -v' to see if everything is correct."
# if $OSTYPE
fi 


installer_CENTOS.sh
#!/bin/bash

# shell script allowing to install all CRISPRCasFinder.pl-v4.3's dependencies on CentOS 
# (could be potentially adapted for Red Hat)
#
# please note that this script requires conda to be installed on your system 
# for further information on conda or bioconda, visit following links:
#   - https://docs.conda.io/en/latest/miniconda.html
#   - http://www.ddocent.com//bioconda/
#
# after running this script ('bash installer_CENTOS.sh'), if your command prompt changes, you may need to type 'exit' command
#
# same version than CRISPRCasFinder.pl, here 4.3.2
# authors: David Couvin, Fabrice Leclerc, Claire Toffano-Nioche

#------------------------
function launchInstall {
# $1 $packageManagmentInstall
# $2 package name
# $3 $LOGFILE
    echo "Installation of $2" >> $3
    $1 $2 >> $3
}
#------------------------

#save the current directory:
CURDIR=`pwd` 

#create log file:
LOGFILE=$CURDIR/install.log
if [ -e $LOGFILE ]; then rm $LOGFILE ; fi

sudo chmod 755 .
sudo chmod 755 *

#create src & bin folders:
echo "create $CURDIR/src and $CURDIR/bin folders" >> $LOGFILE
if [ ! -d $CURDIR/src ];then mkdir $CURDIR/src; fi
if [ ! -d $CURDIR/bin ];then mkdir $CURDIR/bin; fi

#geting OS info: 
ostype=`echo $OSTYPE`
echo "$ostype" >> $LOGFILE

if [ ! -x "$(command -v conda)" ];
then
    echo "Please note that conda (https://docs.conda.io/en/latest/miniconda.html ; http://www.ddocent.com//bioconda/) must be installed on your system."
    echo ""
    echo "You can type following commands to quickly install conda:"
    echo "wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh"
    echo "bash Miniconda3-latest-Linux-x86_64.sh"
	echo "export PATH=/path/to/miniconda3/bin/:$PATH"
	echo "source ~/.bashrc"
    echo "conda init"
    echo "conda config --add channels defaults"
    echo "conda config --add channels bioconda"
    echo "conda config --add channels conda-forge"
	echo ""
	echo ""
	echo "Please retry after the installation of conda!"
	exit 1
fi


if [ ! "$ostype" = "linux-gnu" ]; then
    echo 'Sorry, install process only for OSTYPE linux-gnu (based on yum)'
else
    echo "yum update/upgrade" >> $LOGFILE
    #sudo yum -y update >> $LOGFILE
    #sudo yum -y upgrade >> $LOGFILE
    packageManagmentInstall='sudo yum -y install '
    distribution=''

    if $(uname -m | grep '64'); then
       distribution='Linux_x86_64-64bit'
       echo "$ostype distribution is: 64-bit" >> $LOGFILE
    else
       distribution='Linux_i386-32bit'
       echo "$ostype distribution is: 32-bit" >> $LOGFILE
    fi

    #important packages
    launchInstall "$packageManagmentInstall" "wget" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "curl" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "git" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "java" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "python-unversioned-command" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "parallel" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-App-cpanminus" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-CPAN" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-Module-Build" "$LOGFILE"
  
    #"bioinfo" packages
    launchInstall "$packageManagmentInstall" "hmmer" "$LOGFILE"
    #launchInstall "$packageManagmentInstall" "hmmer-devel" "$LOGFILE"   # may not be necessary
    launchInstall "$packageManagmentInstall" "EMBOSS" "$LOGFILE"  
    #launchInstall "$packageManagmentInstall" "EMBOSS-devel" "$LOGFILE"  # may not be necessary
    launchInstall "$packageManagmentInstall" "EMBOSS-libs" "$LOGFILE"
	
        if [ ! -x "$(command -v fuzznuc)" ] && [ ! -x "$(command -v needle)" ]
	then
		echo "Installation of EMBOSS using Bioconda." >> $LOGFILE
		conda install -c bioconda emboss
	else
		echo "EMBOSS is already installed." >> $LOGFILE
        fi
	
    launchInstall "$packageManagmentInstall" "ncbi-blast+" "$LOGFILE"
	
	if [ ! -x "$(command -v hmmsearch)" ]
	then
		conda install -c biocore hmmer
	else
		echo "HMMER is already installed." >> $LOGFILE
	fi
	
	if [ ! -x "$(command -v makeblastdb)" ]
        then
		wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.9.0/ncbi-blast-2.9.0+-x64-linux.tar.gz
		tar -xzvf ncbi-blast-2.9.0+-x64-linux.tar.gz
		sudo cp ncbi-blast-2.9.0+/bin/* $CURDIR/bin/
	else
		echo "ncbi-blast+ is already installed." >> $LOGFILE
	fi
    
    launchInstall "$packageManagmentInstall" "perl-Time-Piece" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-XML-Simple" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "perl-Digest-MD5" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "gcc" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "gcc-c++" "$LOGFILE"
    launchInstall "$packageManagmentInstall" "make" "$LOGFILE"
    
    #cpanm BioPerl
    #sudo cpanm Unix::Sysexits >> $LOGFILE
    sudo cpanm Bio::Perl >> $LOGFILE

    #BioPerl-Run
    # wget https://cpan.metacpan.org/authors/id/C/CJ/CJFIELDS/BioPerl-Run-1.007003.tar.gz 
    # most recent version of BioPerl-Run will not work. We should use release-1-7-0
    wget https://github.com/bioperl/bioperl-run/archive/release-1-7-0.tar.gz >> $LOGFILE
    tar -xzvf release-1-7-0.tar.gz >> $LOGFILE
    cd bioperl-run-release-1-7-0
    perl Build.PL
    sudo ./Build test
    sudo ./Build install
    cd $CURDIR

    #cpanm Others
    sudo cpanm Bio::FeatureIO >> $LOGFILE
    #sudo cpanm Try::Tiny >> $LOGFILE
    #sudo cpanm Test::Most >> $LOGFILE
    #sudo cpanm JSON::Parse >> $LOGFILE
    sudo cpanm Class::Struct >> $LOGFILE
    sudo cpanm Bio::DB::Fasta >> $LOGFILE
    sudo cpanm File::Copy  >> $LOGFILE
    sudo cpanm Bio::Seq Bio::SeqIO >> $LOGFILE
    #sudo cpanm --force Bio::Tools::Run::Alignment::Muscle >> $LOGFILE
    #sudo cpanm Date::Calc >> $LOGFILE
    sudo cpanm Bio::AlignIO >> $LOGFILE
    sudo cpanm Getopt::Long >> $LOGFILE

    #install vmatch
    if [ ! -x "$(command -v vmatch2)" ] && [ ! -x "$(command -v mkvtree2)" ] && [ ! -x "$(command -v vsubseqselect2)" ]
    then
      echo "Installation of Vmatch" >> $LOGFILE
      echo "change directory to $CURDIR/src" >> $LOGFILE
      cd $CURDIR/src
      wget http://vmatch.de/distributions/vmatch-2.3.0-${distribution}.tar.gz >> $LOGFILE
      tar -zxf vmatch-2.3.0-${distribution}.tar.gz >> $LOGFILE
      gcc -Wall -Werror -fPIC -O3 -shared vmatch-2.3.0-${distribution}/SELECT/sel392.c -o $CURDIR/sel392v2.so >> $LOGFILE
      echo "copy $CURDIR/src/vmatch-2.3.0-${distribution}/vmatch, mkvtree and vsubseqselect to $CURDIR/bin/" >> $LOGFILE
      sudo cp $CURDIR/src/vmatch-2.3.0-${distribution}/vmatch $CURDIR/bin/vmatch
      sudo cp $CURDIR/src/vmatch-2.3.0-${distribution}/mkvtree $CURDIR/bin/mkvtree
      sudo cp $CURDIR/src/vmatch-2.3.0-${distribution}/vsubseqselect $CURDIR/bin/vsubseqselect
      echo "change directory to $CURDIR" >> $LOGFILE
      cd $CURDIR
    else
        echo "Vmatch2, mkvtree2, and vsubseqselect2 are already installed." >> $LOGFILE
    fi

    #install muscle
    echo "Installation of Muscle" >> $LOGFILE
    launchInstall "$packageManagmentInstall" "muscle" "$LOGFILE" # may not work properly
    
    if [ ! -x "$(command -v muscle)" ]; then
      echo "Installation of Muscle (second trial)" >> $LOGFILE
      if [ $distribution == 'Linux_x86_64-64bit' ]
      then
        wget http://drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux64.tar.gz >> $LOGFILE
        tar -xzvf muscle3.8.31_i86linux64.tar.gz >> $LOGFILE
        sudo chmod +x muscle3.8.31_i86linux64
        sudo cp muscle3.8.31_i86linux64 /usr/bin/muscle
      else 
        wget http://drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux32.tar.gz >> $LOGFILE
        tar -xzvf muscle3.8.31_i86linux32.tar.gz >> $LOGFILE
        sudo chmod +x muscle3.8.31_i86linux32
        sudo cp muscle3.8.31_i86linux32 /usr/bin/muscle
      fi
    else
        echo "Muscle is already installed." >> $LOGFILE
    fi

    #install prodigal
    if [ ! -x "$(command -v prodigal)" ] 
    then
        echo "Installation of Prodigal" >> $LOGFILE
        wget https://github.com/hyattpd/Prodigal/archive/GoogleImport.zip >> $LOGFILE
        unzip GoogleImport.zip
        cd Prodigal-GoogleImport
        sudo make install
    else
        echo "Prodigal is already installed." >> $LOGFILE
    fi

    #install macsyfinder
    if [ ! -x "$(command -v macsyfinder)" ] 
    then
      echo "Installation of MacSyFinder using conda/mamba" >> $LOGFILE
      conda install -c conda-forge mamba
      mamba install -c bioconda macsyfinder=2.0
      macsydata install -u CASFinder==3.1.0
      conda install -c bioconda perl-bioperl-core
      #cd ${CURDIR}
      #wget https://github.com/gem-pasteur/macsyfinder/archive/refs/tags/macsyfinder-1.0.5.tar.gz >> $LOGFILE
      #tar -xzf macsyfinder-1.0.5.tar.gz
      #test -d bin ||  mkdir bin
      #cd bin
      #ln -s ../macsyfinder-macsyfinder-1.0.5/bin/macsyfinder
      #cd ${CURDIR}
      #echo "add definition of MACSY_HOME (${CURDIR}/macsyfinder-1.0.5/) in .bashrc" >> $LOGFILE
      #echo "export MACSY_HOME=${CURDIR}/macsyfinder-1.0.5/" >> $HOME/.bashrc
      #echo "add bin folder ($CURDIR/bin) to the definition of PATH in $HOME/.bashrc" >> $LOGFILE
      #echo "export PATH=${CURDIR}/bin:${PATH}" >> $HOME/.bashrc
    else
      echo "MacSyFinder is already installed." >> $LOGFILE
    fi

    #set environment variables
    #source $HOME/.bashrc #this command must be typed directly by user
    
    echo "Installation done."
	echo "You may need to reinstall some Perl's modules (with command: sudo cpanm ...)"
	echo "Please type command 'source ~/.bashrc' in order to reload your .bashrc file, then run 'perl CRISPRCasFinder.pl -v' to see if everything is correct."
# if $OSTYPE
fi 
